diffprivlib/mechanisms/snapping.py

```diff
@@ -0,0 +1,232 @@
+        return True
+
+    def _scale_bound(self):
+        """
+        Scales the lower and upper bounds to be proportionate to sensitivity 1, and symmetrical about 0.
+        For sensitivity 0, only centres the bound, as scaling up and down is not defined.
+
+        Returns
+        -------
+        float
+            A symmetric bound around 0 scaled to sensitivity 1
+        """
+        if self.sensitivity == 0:
+            return (self.upper - self.lower) / 2.0
+        return (self.upper - self.lower) / 2.0 / self.sensitivity
+
+    def _truncate(self, value):
+        if value > self._bound:
+            return self._bound
+        if value < -self._bound:
+            return -self._bound
+
+        return value
+
+    def bias(self, value):
+        raise NotImplementedError
+
+    def variance(self, value):
+        raise NotImplementedError
+
+    def effective_epsilon(self):
+        r"""
+        Returns the effective value used in the Snapping mechanism to give the required :math:`(\epsilon, 0)`-DP, based
+        on the bounds and the machine epsilon.
+        Based on section 5.2 of [Mir12]_.
+
+        Returns
+        -------
+        float
+            The effective value of :math:`\epsilon`
+        """
+        try:
+            return self._epsilon_0
+        except AttributeError:
+            machine_epsilon = np.finfo(float).epsneg
+            return (self.epsilon - 2*machine_epsilon) / (1 + 12*self._bound*machine_epsilon)
+
+    def _scale_and_offset_value(self, value):
+        """
+        Centre value around 0 with symmetric bound and scale to sensitivity 1
+
+        Parameters
+        ----------
+        value : float
+            value to be scaled
+        Returns
+        -------
+        float
+            value offset to be centered on 0 and scaled to sensitivity 1
+        """
+        value_scaled = value / self.sensitivity
+        return value_scaled - self._bound - (self.lower / self.sensitivity)
+
+    def _reverse_scale_and_offset_value(self, value):
+        return (value + self._bound) * self.sensitivity + self.lower
+
+    @staticmethod
+    def _get_nearest_power_of_2(x):
+        def float_to_bits(d):
+            s = struct.pack('>d', d)
+            return struct.unpack('>q', s)[0]
+
+        def bits_to_float(b):
+            s = struct.pack('>q', b)
+            return struct.unpack('>d', s)[0]
+
+        bits = float_to_bits(x)
+        if bits % (1 << 52) == 0:
+            return x
+        return bits_to_float(((bits >> 52) + 1) << 52)
+
+    def _round_to_nearest_power_of_2(self, value):
+        """ Performs the rounding step from [Mir12]_ with ties resolved towards +∞
+
+        Parameters
+        ----------
+        value : float
+            Value to be rounded
+
+        Returns
+        -------
+        float
+            Rounded value
+
+        """
+        if self._epsilon_0 == float('inf'):  # infinitely small rounding
```

Comment:
`_epsilon_0` is only infinite when `epsilon` is.
```suggestion
        if self.epsilon == float('inf'):  # infinitely small rounding
```